### PR TITLE
Ubiquiti Unifi Parser & Libs

### PR DESCRIPTION
**Pull request by h00die**
This PR overhauls the Ubiuiti Unifi system to:
 1) add a new module to simply upload a zip of the backup (if you got it some other way than what MSF provides)
2) now works on UDM-Pro
3) Now, after the ZIP is repaired, we extract the mongo-db, and pull out all the juicy bits.

### What juicy bits?

We also print if its enabled or disabled, because that's useful information!

Database:
1) Creds: username, email, and password hash of admin
2) Creds: radius server and secret
3) Creds: mesh wifi name and password (not hashed)
4) Creds: ssh username, hash, and password
5) Creds: snmp v2 and v3 passwords
6) Creds: wifi names, encryption types, and passwords
7) Hosts: add all the unifi controled devices by IP, model, and name
8) Hosts: add all devices by IP, mac, name, and any info field

Print:

1) firewall rules
2) firewall groups
 

## Verification

- [ ] Try uploading a unifi zip to `auxiliary/admin/ubiquiti/ubiquiti_config`
- [ ] Get a shell (SSH is fine) and try the post module to do the same
- [ ] run spec to make sure tests are all good

### CODE REVIEW

**Comment by h00die** [comment]
Sample output (heavily redacted)
```
resource (ubnt.rb)> use auxiliary/admin/ubiquiti/ubiquiti_config
resource (ubnt.rb)> set config '/root/5.10.26-20190822-2205.unf'
config => /root/5.10.26-20190822-2205.unf
resource (ubnt.rb)> set rhosts 127.0.0.1
rhosts => 127.0.0.1
resource (ubnt.rb)> run
[*] Running module against 127.0.0.1
[+] File DECRYPTED.  Still needs to be repaired
[*] Attempting to repair zip file (this is normal and takes some time)
[+] File DECRYPTED and REPAIRED and saved to /tmp/fixed_zip.zip20200111-28420-1vccznk.
[*] extracting db.gz
[*] Converting config BSON to JSON
[+] Admin user adminusername with email admin@admin.com found with password hash $6$<redacted>.
[+] Radius server: 1.1.1.1:1812 with secret 'radiussecret'
[+] Mesh Wifi Network vwire-<redacted> password <redacted>
[+] SSH user admin found with password <redacted> and hash $6$<redacted>
[+] SNMP v2 enabled with password public
[+] SNMP v3 enabled with username usernamesnmpv3 password passwordsnmpv3
[*] Enabled Firewall Rule '<redacted>': reject src group: <redacted> protocol: all
[+] Enabled wifi <redacted> on wpapsk(wpa2,ccmp) has password <redacted>
[*] Firewall Group: <redacted>, group type: address-group, members: 1.1.1.1, 2.2.2.2
[+] Unifi Device USG of model UGW3 on 4.4.4.4
[+] Network Device desktop (<redacted>) found
[+] Network Device  (<redacted>) on IP 9.9.9.9 with name <redacted> found
```

**Comment by h00die** [comment]
UDM-Pro keeps adding features.  Once this lands, i'll check for some more stuff like their IDS/IPS feature, GEoIP filtering, and Network Scanners.
If the network scanner data is in the database, it'll be like importing a free nmap scan!

**Comment by h00die** [comment]
pending automated testing, this is good to go for testing now!

**Comment by h00die** [comment]
Updated with https://github.com/h00die/metasploit-framework/issues/13#issuecomment-573357213

**Comment by h00die** [comment]
TBH, not sure how to patch in the spec to avoid this.  Tried several ways with adding hardcoded `core`s to the expect and `login_data`. None seem to work

**Comment by smcintyre-r7** [comment]
I pulled this down and tried to test it and I got the following stack trace:

```
msf5 auxiliary(admin/ubiquiti/ubiquiti_config) > set CONFIG /tmp/5.12.35-20200221-1511.unf
CONFIG => /tmp/5.12.35-20200221-1511.unf
msf5 auxiliary(admin/ubiquiti/ubiquiti_config) > set RHOSTS 192.168.250.6
RHOSTS => 192.168.250.6
msf5 auxiliary(admin/ubiquiti/ubiquiti_config) > run
[*] Running module against 192.168.250.6

[+] File DECRYPTED.  Still needs to be repaired
[*] Attempting to repair zip file (this is normal and takes some time)
[+] File DECRYPTED and REPAIRED and saved to /tmp/fixed_zip.zip20200221-1146003-4beulz.
[*] extracting db.gz
[*] Converting config BSON to JSON
[-] Auxiliary failed: NameError uninitialized constant Metasploit::Credential::Creation::ApplicationRecord
[-] Call stack:
[-]   /home/smcintyre/.rvm/gems/ruby-2.6.2@metasploit-framework/gems/metasploit-credential-4.0.0/lib/metasploit/credential/creation.rb:11:in `active_db?'
[-]   /home/smcintyre/.rvm/gems/ruby-2.6.2@metasploit-framework/gems/metasploit-credential-4.0.0/lib/metasploit/credential/creation.rb:110:in `create_credential'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/data_service/proxy/credential_data_proxy.rb:41:in `block in create_credential_and_login'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/data_service/proxy/core.rb:166:in `data_service_operation'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/metasploit/framework/data_service/proxy/credential_data_proxy.rb:39:in `create_credential_and_login'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/report.rb:54:in `create_credential_and_login'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/ubiquiti.rb:214:in `block in process_settings'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/ubiquiti.rb:185:in `each'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/ubiquiti.rb:185:in `process_settings'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/ubiquiti.rb:320:in `block in unifi_config_eater'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/ubiquiti.rb:313:in `each'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/msf/core/auxiliary/ubiquiti.rb:313:in `unifi_config_eater'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb:91:in `run'
[*] Auxiliary module execution completed
msf5 auxiliary(admin/ubiquiti/ubiquiti_config) > db_status 
[*] Connected to metasploit. Connection type: postgresql.
msf5 auxiliary(admin/ubiquiti/ubiquiti_config) >
```

I haven't taken a look through the code too thoroughly yet though.

**Comment by h00die** [comment]
You're running a newer creds
```
metasploit-credential-4.0.0
```
vs
```
metasploit-framework# bundle install | grep cred
Using metasploit-credential 3.0.4
```
I'm not hitting that error on my import file, even w/ a rebase.  Don't remember ever seeing that variable, or setting it (in all my creds exp).  Maybe its a bug in -credential 4 gem builds?

File: modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb
Comment (wvu-r7) [inline modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb]: This is my contribution.

```suggestion
      'Author'        => ['h00die'],
```

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (smcintyre-r7) [inline lib/msf/core/auxiliary/ubiquiti.rb]: The `nil` here isn't necessary. The whole `else` block could actually be removed since the conditional block above it always returns.

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (smcintyre-r7) [inline lib/msf/core/auxiliary/ubiquiti.rb]: You could bump these 4 lines out of this block and up a level so they don't need to be repeated when the line is determined to be an SNMPv3 entry.

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (smcintyre-r7) [inline lib/msf/core/auxiliary/ubiquiti.rb]: 5 out of 7 times you use this template (via `credential_data.dup`) you set `cred[:private_type] = :password`. Why not just set it here and continue to override it as necessary when the type is a hash.

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (smcintyre-r7) [inline lib/msf/core/auxiliary/ubiquiti.rb]: I see you're filtering out the default, public NTP server. :+1: :100:

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (smcintyre-r7) [inline lib/msf/core/auxiliary/ubiquiti.rb]: You could make this quite a bit more dynamic by using respond_to and send here.
The whole case block could reduced to
```ruby
next unless self.respond_to? "process_#{key}"
self.send("process_#{key}", value)
```

`process_user_device` would need to be switched to `process_user` and you'd need an alternative way to ingest the `credential_data` hash, either a function would work (you wouldn't have to `.dup` it any more) or using a constant `CREDENTIAL_TEMPLATE`.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
I pulled this down and tested it again. The only way I could get it to work though was to rebase it on master to pull in the latest Gemfile changes. It rebased cleanly, would you be able to do that and update this branch?

File: modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb
Comment (h00die) [reply modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb]: good suggestion.  Believe it was a bit of copy pasta from Juniper where there were two actions, however this will bring it inline with Brocade

File: modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb
Comment (h00die) [reply modules/auxiliary/admin/ubiquiti/ubiquiti_config.rb]: yup good call on both parts

**Comment by h00die** [comment]
some updates, not finished yet, just low on time ATM.

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (h00die) [reply lib/msf/core/auxiliary/ubiquiti.rb]: I'm not opposed to cleaning this up, but this is based off of the cisco module. Brocade, Juniper, and Unifi are all based on the Cisco module, so I'd rather hold off on this change and do a diff PR which keeps all of those in line and upgrades them all at once.

File: lib/msf/core/auxiliary/ubiquiti.rb
Comment (h00die) [reply lib/msf/core/auxiliary/ubiquiti.rb]: Same as comment below.  This is a good point, but i'd like to evaluate it in brocade/juniper/cisco as well and just do the change to all the modules at once for consistency

**Comment by h00die** [comment]
I tested this against my UDM-Pro as a post module, and the aux using a unf.  All working for me.
UDM-Pro is getting a LOT of feature quickly, A new one is a network discovery feature where it seems to run nmap and store the results.  I'll add processing for this in a future PR since this is already a bunch of changes.

**Review by zeroSteiner — APPROVED** [review]
Tested successfully and all comments have been addressed! I'll get this landed momentarily.

(Not posting output since it's full of information regarding my network that'd need to be censored)

**Comment by zeroSteiner** [comment]
## Release Notes
This adds a new mixin for importing useful information from Ubiquiti UniFi backup files. The new mixin is then also used by a new auxiliary file that can ingest files collected through arbitrary means.

**Pull request merged**
